contrib/upgradability-patterns.asciidoc

```diff
@@ -0,0 +1,367 @@
+   * @param record sha3 key 
+   */
+  function deleteBooleanValue(bytes32 record) public onlyOwner {
+    delete BooleanValues[record];
+  }
+}
+----
+
+*CounterContractV1*
+----
+pragma solidity ^0.4.13;
+
+import "zeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "EternalStorage.sol";
+
+contract CounterWithEternalStorage is Pausable {
+
+  EternalStorage eternalStorage;
+
+  function CounterWithEternalStorage() {
+    eternalStorage = new EternalStorage();
+  }
+
+  function getCount() public constant returns(uint256) {
+    return eternalStorage.UIntValues(keccak256("counter"));
+  } 
+    
+  function increment() public whenNotPaused returns (uint) {
+    uint count = getCount();
+    uint value = count.add(1);
+    eternalStorage.setUIntValue(keccak256("counter"), value);
+    return value;
+  }
+
+  function transferEternalStorageOwnership(address recipient) public onlyOwner whenNotPaused {
+    Ownable(eternalStorage).transferOwnership(recipient);
+  }
+}
+----
+
+So let's say for some reason you want to upgrade the *CounterContractV1* so that `increment()` increase the counter by 2; you just need to deploy a *CounterContractV2* that point to the same *EternalStorage* instead of creating it in the constructor. Then you call `transferEternalStorageOwnership(AddressOfCounterContractV2)` on *CounterContractV1* with the address of *CounterContractV2* as argument.
+
+*CounterContractV2*
+----
+pragma solidity ^0.4.13;
+
+import "zeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "EternalStorage.sol";
+
+contract CounterWithEternalStorage is Pausable {
+
+  EternalStorage eternalStorage;
+
+  function CounterWithEternalStorage(address _eternalStorage) {
+    eternalStorage = EternalStorage(_eternalStorage);
+  }
+
+  function getCount() public constant returns(uint256) {
+    return eternalStorage.UIntValues(keccak256("counter"));
+  } 
+    
+  function increment() public whenNotPaused returns (uint) {
+    uint count = getCount();
+    uint value = count.add(2);
+    eternalStorage.setUIntValue(keccak256("counter"), value);
+    return value;
+  }
+
+  function transferEternalStorageOwnership(address recipient) public onlyOwner whenNotPaused {
+    Ownable(eternalStorage).transferOwnership(recipient);
+  }
+}
+----
+
+
+[[proxy_libraries]]
+=== Proxy Libraries
+A proxy libraries allow for a contract to maintains the same address, and the same storage while changing the logic inside of it.
+This is done through the use of 4 contracts the *main contract*, the *dispatcher*, the *dispatcher storage* and the library that is the upgradeable part.
+
+advantages:
+
+* The address of the contract never change
+
+drawbacks:
+
+* More complex to implement
+* Need to use Assembly
+* You need to keep the same storage footprint
+
+image::images/proxy-lib.png["Proxy-libraries"]
+
+==== Implementation
+
+*MainContract*
+----
+pragma solidity ^0.4.8;
+
+import "./LibInterface.sol";
```

Comment:
Would recommend moving the `LibInterface.sol` and `DispatcherStorage.sol` to the top of this section because they are being referenced by the client code which would make it easier for the reader to follow through.

Reply:
I don't understand what you mean by "client code"